### ISSUE TITLE
4.7 regression: Array.flat(infinity) leads to "Type instantiation is excessively deep and possibly infinite." error

### ISSUE THREAD

**Issue by peterjuras** [microsoft/TypeScript#49280]
# Bug Report

### 🔎 Search Terms

- Array.flat
- Type instantiation is excessively deep and possibly infinite.
- Nested arrays

### 🕗 Version & Regression Information

- This changed between versions 4.6 and 4.7

### ⏯ Playground Link

[TypeScript Workbench with relevant code](https://www.typescriptlang.org/dev/bug-workbench/?target=6&ts=4.7.0-beta#code/PTAEAEBcEMCcHMCmkBcoCiBlATABgIwCcAUMQJYB2kisAZtAMaKgDCA9hbWfKAN7GhQAB1hshAfjQBnSLErwA3MQC+pStTqNm6AB7UKAE0QH2nbqER7Ehqaw5cevVcUgBPIcwByiGcdMPbAF5QAEFYWGhXAB5dfSMTe3MAH1BvXwSzeCkAPiViBg4ZUALMqTQ06gyAgG0AXVBgurySotoAG2hIfWMG4sSsgDp2zoAKAEkzCjI3AEolIA)

### 💻 Code

```ts
interface Config {
  prop?: string;
}

interface ExtendedConfig extends Config {}

type NestedConfigs = Array<ExtendedConfig | NestedConfigs>;

const configs: NestedConfigs[] = [];

const flattened = configs.flat(Infinity);
```

### 🙁 Actual behavior

TypeScript raises the following error for the `configs.flat(Infinity)` call:

`Type instantiation is excessively deep and possibly infinite. ts(2589)`

### 🙂 Expected behavior

No error should be raised.

**Comment by grant-dennison** [comment]
Ran into this today as well. 4.6 typed the output as `any[]`. ([playground](https://www.typescriptlang.org/play?target=6&ts=4.6.4#code/C4TwDgpgBAYg9nKBeKBvKAjAhgJwFxQDOwOAlgHYDmUAvgLABQokUAIhBGAII45YgAeACoA+ZFAAUQqAB82Hbr37CRASgDaAXUbNoPPiADyOAMoVKAGwgrx0ue077loxowAmEAMYXc0T3HJiKA9OAicjU3MrAXg4EUZ-QOAoADMfZJR1ELBNADo0rGAJAElyFIpSUFVGAHoaqAaoAD0AfkYgA))

**Comment by sepiette** [comment]
also ran into this in an upgrade from 4.6.4 to 4.7.2

**Comment by grant-dennison** [comment]
I'm not sure what I think the correct type of the output is, but it does seem to me that the TypeScript compiler should allow `.flat(Infinity)` (without type assertion) without straight-up failing the compile since it is valid and common (e.g. [example on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat#flattening_nested_arrays)) JavaScript.

Maybe it would make sense to special-case `Infinity` to wind up with `unknown[]` or `any[]`?

**Comment by tjjfvi** [comment]
This could be addressed if a literal type for `Infinity` was added: [playground link using `1e309`](https://www.typescriptlang.org/play?target=6#code/C4TwDgpgBAYgNgQ2AQQE6oSAkgOwGYCWOBoAPACoB8UAvAFBRQDa5AulAD7NtQQAewCDgAmAZ2ZE8EVFACqrBoygB+ObwFCxUAEoQEwgPY44INBhClJ0qAFFKipUtUsANLfb9BI8Uxtu2Do4qUDgQAG7SgY4AXLCIKOiYuITEZHZRSrGygbGhEagA3HR0oJBxSGaYAHIArgC2AEbSFNQ0zIpc5B1QABQs7Fz96l5aTFYy8hmqssOa4rr6RiaVFuO29kFO3G42HhrezH5QAZuOqnmRp5nlCea1jc3pV7nhl0Ev+QCUdKxMAAysIolcDQeAVRIgADCRlEwFwwFIZjcABEIGBgAALWYHHD1JqoVqKVHorGeOZQAC0AEZAqozDkoGZsVpUHpDMYQL01rhQqgzJ8mApNqowbdMNCcLD4aQMowedIkbLmNS3H83FS3AAmNwAZjcABY3ABWNwANjcAHY3AAONwATnVaqgVI1zu1zr1zsNzpNzvNzqtzttzodUE1AKYxMxQtOG3ejPQQNKoPiK0R6BRaMxzPEuIeBNoijz+JzUCjGNpNxW93x6YJDPLpapEB1fztUyrEOSRBIFjMcZinfMEqlOARSLLWYxlCBwggAGNEKyoABzOAGBoIOBQADeiiIglQeAQ8+gaaou8CeHimsRKNLxesbSplB6mIIoliyDcc5JyliyKfLEoppt+ZaUIKRSMAAvnQsF0Ae0jHqeUASoQK6XowYCoAYYD-lAsKoEQK5FPBiFHie0A2Psc7CGhBAYWSBz0RhO7wcmUBVBAsIQHRRjoeIbRptRIy8SxnCcdxgh8fgDGiDOxRzouCDLvOMLAFAamySun6STxMkCZBdBKUu0BabCUDADUYBwBAsRMK6TAAEQIE5bg7lADSxE585OVA0GsG4gpBUwoVOcITmsFFUVFCZKlmepIR4tIuTJYUxTmRpWkCVStCafxckAHTXkgmo9FSnxFAA9FVjgAHrKHQmX5dpohUn8eXZUVJXAGV7WVXQNX1Y1zVdTp3apJybRjaIxU3uVLZtgNQ1KA1TWJTNNZPi1AlzaVPSPqgy21atI2JVZNkQLlbQXbZe29eVx3DetkoabdECanl733WVmpPadL0We9E29l91l3T1fWLXa-2MGtzXvVtMg3eDEA-QdaWw1Aa10PwYAGKgGlsUUQA)

**Comment by tadhgmister** [comment]
This is solvable now without infinity literal. @tjjfvi unfortunately your code only checks direct recursive references not recursive loops, [Playground Link][1]

```ts
type A = number | B[]
type B = string | A[]

declare const configsBreaking: A[]

const configsInfinity2 = configsBreaking.flat2(1e309);
/// ERROR:  Type instantiation is excessively deep and possibly infinite.(2589)
```

Much like a runtime algorithm, the way to prevent the type system from infinitely recusing is by tracking what data structures you've already seen and short circuiting if they are self referencing.

This does work: ([Playground Link including tests][2])

```ts
/** 
 * Counter in type context, 
 * `Counter.Value<Counter.ZERO>` gives 0
 * `Counter.Value<Counter.Inc<Counter.ZERO>>` gives 1 etc
 */
type Counter = Counter.Valid;
namespace Counter {
  /** constraint for type context counters, same as top level `Counter` type. */ 
  export interface Valid { [k]: e[] }
  declare const k: unique symbol;
  type e = 'unused dummy';

  /** count of zero */
  export type ZERO = Record<typeof k,[]>
  /** extracts the number literal from a counter */
  export type Value<C extends Counter> = C[typeof k]["length"]
  /** increments a counter to one higher */
  export type Inc<C extends Counter> = {[k]: [e, ...C[typeof k]]}
}

/** replaces array types with their element type, leaves non array types as is */
type UnwrapArray<T> = T extends ReadonlyArray<infer E> ? E : T
/** returns an array type that corresponds to flattening one layer of the list */
type UnwrapOneLayerTuple<
    T extends any[]
> = T extends [infer A, ...infer R]
    // if T is a tuple type handle just the first element and recurse to rest of tuple
    ? A extends ReadonlyArray<any> 
      ? [...A, ...UnwrapOneLayerTuple<R>]
      : [A, ...UnwrapOneLayerTuple<R>]
    // if T is not a tuple type use a homomorphic transformation on the elements
    // generally it maps [] to [] and (A|B)[] to (UnwrapArray<A>|UnwrapArray<B>)[]
    : {[K in keyof T]: UnwrapArray<T[K]>}

type FlatArray<
  T extends unknown[], 
  Depth extends number, 
  /** counter for how many steps down the depth we have traversed */
  C extends Counter = Counter.ZERO,
  /** all possible list states visited so far, if adding a new step doesn't change this we short circuit. */ 
  AllSteps extends unknown[] = T,
  /** the next step, effectively to cache it. Typescript is probably smart enough to reuse this computation 
   * if present multiple times in the definition but idk how the internals work 
   * so being here ensures it computes consistently based on the input T every time.
   */
  NextStep extends UnwrapOneLayerTuple<T> = UnwrapOneLayerTuple<T>,
> = [NextStep, T] extends [T, NextStep]
  // T (current flattening status) and NextStep are assignable to each other
  // so either it is finitely nested structure that has been fully flattened
  // or there is single step recursion like `type P = (1 | P)[]`
  // Since T and NextStep are resolved differently it is not obvious which would be more useful
  // to return in this case, I've included both with debugging info for your own reference.
    ? (T|NextStep) & DEBUG<"no change at finite depth", {depth:Counter.Value<C>, T:T, N:NextStep}>
       
  : [AllSteps|NextStep] extends [AllSteps]
  // we have some form of recursive loop, such as `type A = 1 | B[]; type B = 2 | A[]`
  // such that unwrapping more layers results in the same possible types we've already seen
  // just give basic array type of the viable element types
    ? (AllSteps[number])[]  & DEBUG<"loop recursion", Counter.Value<C>> 
    
  : (Counter.Value<C> extends Depth ? T : never) // add T if the depth parameter contains the current depth
  // and union the possible result from this depth with the result from the next depth, 
  // as long as there is a next depth to deal with.
  | (Exclude<Depth, Counter.Value<C>> extends never ? never 
      : FlatArray<NextStep, 
                  Exclude<Depth, Counter.Value<C>>,
                  Counter.Inc<C>,
                  AllSteps|NextStep,
                  UnwrapOneLayerTuple<NextStep>
                  >)
```

This has worked reasonably on all the cases I've thrown at it, I feel like there should be a way to define a type that gets more convoluted as you look deeper in its recursive definition but not sure how to actually construct it to see this throw an 'infinite recursion' error.  The one place this does fall short is that negative or fractional depths are handled reasonably by the runtime but treated as "large enough we stopped getting new results from more flattening" by this type definition, I considered using string templates to give `unknown[]` in those cases but cases like `1.3e100` and `120e-1` would pose issues for that approach.

Another problem that is totally solvable is that passing a `number` gives a union of all unique depths where as an arbitrarily big number gives only the deepest option so `Infinity` being typed as `number` and not a literal is annoying. If it was an identify-able literal we could even strip out array type elements once we stabilize but doing that for recursive types in the same definition that retains tuple-ness in finite structures is tricky. I had a version that did it but it was much harder to test so dropped it while refactoring.

These are the tests I tried organized into `namespace`s so they are easy to collapse in the playground. There were a few more I tried to get a recursion error but they didn't do anything distinct from the behaviour of these tests. ([same link as above][2])

```ts

/** a very large effectively infinite number, but typed as a literal number not 'number' */
const Inf = 1e300;


declare global {
  interface Array<T> {
    flat2<A extends unknown[], D extends number = 1>(this: A, depth?: D): FlatArray<A, D>;
  }
}
namespace SimpleRecursion {
  type P = (1|P)[];
  declare const config: P;
  const res = config.flat2(4 as number);
  //    ^? (P | (1 | P)[]) & DEBUG<"no change at finite depth", {depth: 0; T: P; N: (1 | P)[];}>
  // confirm that is equivelent to original nested structure.
  const sanity_check: P = res;
}

namespace OP_case {
  interface Config {
    prop?: string;
  }
  interface ExtendedConfig extends Config {}
  export type NestedConfigs = Array<ExtendedConfig | NestedConfigs>;
  export declare const config: NestedConfigs[];

  // unsure if the intention was to have a minimum of 2 layers of arrays but can confirm this works.
  const op_flatten_once: NestedConfigs = config.flat2(1);
  const res_op = config.flat2(Inf);
  //    ^?  (ExtendedConfig | NestedConfigs)[] & DEBUG<...>

}
namespace RecursiveLoop {
  type A = number | B[]
  type B = string | A[]
  export declare const config: A[];
  // might be desirable for this to give (number | string)[] instead of (A|B)[]
  // but that would require an infinity literal type.
  // and more convoluted handling to not break tuple cases 
  const res_loop = config.flat2(Inf);
  //    ^? (B | A)[] & DEBUG<"loop recursion", 2>
}
namespace FiniteTuple{
  export declare const config: [
    1, 
    [[[["d"]]]],
    [], 
    [
      "a", 
      { b: "c" }
    ], 
  ];

  const res1 = FiniteTuple.config.flat2();
  //    ^? [1, [[["d"]]], "a", { b: "c"; }]
  const res2 = FiniteTuple.config.flat2(2);
  //    ^? [1, [["d"]], "a",  {b: "c"; }]
  const res3 = FiniteTuple.config.flat2(3);
  //    ^? [1, ["d"], "a",  {b: "c"; }]
  const res4 = FiniteTuple.config.flat2(4);
  //    ^? [1, "d", "a",  {b: "c"; }] 

  const res_inf: typeof res4 = FiniteTuple.config.flat2(Inf);
  type Exp_depth_2_or_3 = typeof res2 | typeof res3;
  const res_union: Exp_depth_2_or_3 = FiniteTuple.config.flat2(2 as (2|3));
  // note `typeof config` is here because [[1]].flat(0) is valid.
  type Exp_any_depth = typeof config | typeof res1 | Exp_depth_2_or_3 | typeof res4
  const res_num: Exp_any_depth = FiniteTuple.config.flat2(Infinity as number);


  /// note that this implementation treats negative or fractional numbers 
  /// as equivelent to arbitrarily big numbers that will fully flatten the array.
  /// I tried using string templates to handle this but 1.3e300 and 100e-1 got treated incorrectly.
  const res_maybe_problematic_1: typeof res_tuple_inf = FiniteTuple.config.flat2(-3);
  const res_maybe_problematic_2: typeof res_tuple_inf = FiniteTuple.config.flat2(1.5);
}

namespace AnyChecks {
  declare const threeD_matrix: any[][][];
  // This case probably should give any[] but it short circuits earlier
  // because of `any` assignability rules.
  const res = threeD_matrix.flat2(3);
  //    ^? (any[][] | any[][][]) & DEBUG<"no change at finite depth", {depth: 0;  T: any[][][]; N: any[][];}>
}
```

I'm not sure if this is sufficiently more complicated than any other types in the standard library that it is undesirable for the complexity, even if we can establish that it works reasonably for every possible case, but this is absolutely doable.

[1]: https://www.typescriptlang.org/play/?target=6#code/C4TwDgpgBAYgNgQ2AQQE6oSAkgOwGYCWOBoAPACoB8UAvAFBRQDa5AulAD7NtQQAewCDgAmAZ2ZE8EVFACqrBoygB+ObwFCxUAEoQEwgPY44INBhClJ0qAFFKipUtUsANLfb9BI8Uxtu2Do4qUDgQAG7SgY4AXLCIKOiYuITEZHZRSrGygbGhEagA3HR0oJBxSGaYAHIArgC2AEbSFNQ0zIpc5B1QABQs7Fz96l5aTFYy8hmqssOa4rr6RiaVFuO29kFO3G42HhrezH5QAZuOqnmRp5nlCea1jc3pV7nhl0Ev+QCUdKxMAAysIolcDQeAVRIgADCRlEwFwwFIZjcABEIGBgAALWYHHD1JqoVqKVHorGeOZQAC0AEZAqozDkoGZsVpUHpDMYQL01rhQqgzJ8mApNqowbdMNCcLD4aQMowedIkbLmNS3H83FS3AAmNwAZjcABY3ABWNwANjcAHY3AAONwATnVaqgVI1zu1zr1zsNzpNzvNzqtzttzodUE1AKYxMxQtOG3ejPQQNKoPiK0R6BRaMxzPEuIeBNoijz+JzUCjGNpNxW93x6YJDPLpapEB1fztUyrEOSRBIFjMcZinfMEqlOARSLLWYxlCBwggAGNEKyoABzOAGBoIOBQADeiiIglQeAQ8+gaaou8CeHimsRKNLxesbSplB6mIIoliyDcc5JyliyKfLEoppt+ZaUIKRSMAAvnQsHAmUyC0CEeLWFwABCgoIdA6HIbCqBECunCMlhdBzouCDLvOMLAFA1H4AQK6iOhrIIAA1oRX6kfRsJ0UYhBMd2qQgJqyH0QJzGsRxOArgAdNeSCaj0zatnanxAvwYAGKgtE7tBRRAA

[2]: https://www.typescriptlang.org/play/?ssl=159&ssc=1&pln=152&pc=1#code/PQKhAIChwhLA7ALgUwE4GdkGNHkQC1nXAHdYDwBDeAewLTwE8AHZPG8AW0tQGtwAJsgBGAVwDm4BADMaqbolg14pQgBs2KdIvjjocaeAAiAUQBCAVQDiAOhMA5AIJmAMianFMuRB2mU1mHiExETgqMjMapRYyAKk5PjgAAai8Ly0JPBJ4OgcBJS4CCgY2DqSZBTUdPgMiCxsAjTIxLS4lNLSpVKI+sCQdazG5tYAPACy6JLIAB4o8AKeiKgI4gA0xgWUALwAbjSwAgB84FtDlrYOzm7gM3MLeKiibAD8Z9Y2Yzy8aOOTq0abY4ALnAqXSNEykHglE4zWY0TYpnO4AA3tBwKAICQaiovOxwH4AmwimhMDgVm8rPFKq0aqgmIN0PgaKI1AJ4AByNodLrkdEwVTIXHIbx5R5sSgCATkJTQtSCTbgLDUcDCNgzZjhdCYOLCRhSJCk0oUpGjABEABVHFYzasABrHCqJSgM5A2fkgPrgG7TZhybz1cCXVzuU5LJ4Abkg6MxXC+tQ42jkDREEkEIsosACMC9Pr9qEKhtQfhi4E+fB+EzWG0QlGOaO93oA2rwALogptV9YA2utqPegC+6KEWCi4SVym04F4INSsAAjk8coxOMIaGoo0PY-pwABhFlFg2uieG2brHdJfepYo2ABq-ieIyvRZsAC0TAAlADyh2y4lgOzNOAAAMF7Pje95qI+4FoDYACS8BYE+B43u+36HL+4D-oBxAAIw3IgWC9P0gYwfSpxkXe-gHFG0Kwug8KlmRqIxmAJ7aKgmZIAScjHlgygoLME7XqS6zoDCErED4zDgBogHypeKFoNkAxujmUDehq-oGsUJZsJBByouALbtjcTatuAQ7eiOY5sPx8BTjOoLwAuS7oCua4buiqk3Cc4Acqkog6oIoicJwjAclGrEQPx17gDQhgAF5oBwnrolpBbHmhX5+R+2ByAIIyqQl06rOZhzRT6SzRIgUk1OA8ChWq9JqOQaD+ASqA0JwVDCYeaWab62k+ZB0FVUK9xkccFFNsVhhtk2ZoaLoBBmq2lUIFg4SwkgxAurFh4+PF8BsIQ4h0jm6VDZlPkIUhu7jfMxBTX5KImR2yDrDY327rN9QlW2rZDkOkCxuEkQIntqCcfqqnEE6QTILA9LIBoO0Bqw6waJQOENcoVDQ5QsP1HtITEGlPkWPAJCccwjiE4wIwWtN4AWo99x5ZKyhqIw9MwyMMgMCYxyvO4IIWqDbHhIgoioA5VAqDwMPHvkuD8dDcLKPcR3SFEiBzBSyhsFEjAMCV9CyUQuAU4GVM05QzBfidLhE2gFqiJEyAjPyrPs3t8CMOZkAs2ztwTcQTaC-Sjhfd9UfgB+62NhiwBSIYbOhC6Mue8e+DUAIGjgAAVkF3j1dIyNTqjyDowrcThFgsuBEdWq4ObHsaD7ryOH7CfIFz8A83zRMjNQjDHD73qvE2302DH4Cz3btNO8gLum6g7ueyMH6HEnyfgB28+L9Ty-O67G8d17O9796wCp7A6ceHjbR4Jfx5BRK4DMpw3VyMwhBYAeNQdAsh5AFFlMdRGNw0ZClqj7O+WEhTtTUDzbocZmARwskdcyddwAAApHAAB8zAAEocFHTwUvB2w9GaOEOIQqhdMGYjDMIcMhN8D6oibAAaSPN8RgJULSmUYTQpmPDWyHBBj5AAYnrUR6JQ6zHDs5cEmRzLnm9EYCIFAw5PQak1NAGiMRsQOsUHi9JmQkDjAHHIKAMGCAhCoC2QhmAUBIKdHGmhOKARKHEAae5e7MQokpVAb5PxflWJVfw8o-TalgMIQurUpzaAKEBHYRA2pxFyASHg6wH5UClBSF0J0rHaAiA45onI1Z510JoYIpA2BMm0lgZGjdyA2HUuiRwKCADKdjiC6PuGCDI8AcGnAtJE2+bELYnSEmU5g6xkA8nJPJWGHBlRYHqu01mJMtqwFcU-TUNBhCUASfqdA3BMpChZOdfE4QP5BFCPxTgzBRC1kUPjD0adwCamaLArgrJFA50UPRI8zilkIBlPjMQhQBD8EsVAkkct-DwzkPwL52S1QUjpOqByssgLkAnC8t5QF7LoCtrA1BJzgr4wtggV5uBQ4+NhrAWE7pGz+PsLcPp5TBnEEYSvNebtL5MxZgKs+69N4aFFZElmTYuWzB5Qs1mFk+XGQmeABViAlV7wQWzPBjcNbcV1gUA2uhbEFCCiQ3BWqlUE0kuS8Q0IEmaA4P3TZ8V6CoBjKnbJSMvVoNCBXFyKBUEnTKVk8MOB8VBAKF-SgxA1RCgJKyVBJr9ZIIED6+K9IvXEk8CsQu8ywjYCbhA1q3xkg+QAAp+TwfhQh4Bq3sKSNmnpm02Bs3zpq7ldj7UltyGoQCcRpQ8nCEgVBhLQitHisIdJLJ4YAMSCQFkbJVRsB-uOD+0hWTZpbiKWWKgECPOIMqTA6w4IckAgaUcoghC6mqNSRIQgxDiH-OamQvheICNlvFTIJbOjjpiOyxsrw8EWkIbaux1qABklIRhmloEqGp4gJS4GDW1dMrj8C2lRC4ggQJKKjS9ruQ46wLRAg1fYIEUGIgDgqvvb06JD69P6ZB3tERVVKL0U2bpaglXoF1andx8br25FhOYnqJUG5luvWoGgNBlXoFEB6hNVbAw91OA28AZhzIRmPGYPyAAmcAjbHDmVbbfX1KnEiq2cvbZgzAKSbuNufYgWpAUhCcfVcSEnYnkpda6eGyAr0SjUOESU5zkBCmzSXKc2E2DUtgIApWRNjzm3quk05hdq61zhl3fBfGBNNkaquNArZ2Hejg6aKwCH5OKZLYa8lyhcNEYfCRjCGlGzMfwW1qCHXe5aOw+AV4bMQQnR8dahBko4gZ0MOC4b8JOKwjMfZWsCA6p2VluO3A+H8DZu7XOWl9V-PxMLh5tQ6Guo9QIKEPbT6oEXau91KBszdvaPwEY6bxB5PmrU3mp+xTbhYYoEdIQHUnQgcbXgkw0xb1CBGENgg6w+vQU62qibDBXiY-pJPThsiCiiNo8qvHjHGOw-h17JHn29whKov1p8GFJlk5Z96Sid1GfM9Z2TorbHidc+5-vcVq9z5Sq9sThjgvGNsOjNGWMLpmWyR4Khm4yzFCrINBhlA+iyuoHWDC10cQ1MulasUDqpXmrP38hbtAHJLpktwAhQwWnkAAGZgLASipAGyPA2DiHkyc+UDYdJoD0uAURzMWLJ3TUZkYPc1XDMceo4wvcbfkXALhQ4eDbvoBBPPPbzwQRGBISCAniBRHzyMIcfsllIBDjonCBE4B20vI0HlJrEDg81rrbhQhza9PDmwLZdiatlAV3ECCatNeHcDr8vZcfNgY94IACxUBaAY1AJCa8IMbAAPWeNm+fyMbt50KAMxcAFq7cSOnIWA77zfNBQJGx40bwggZn+JENjAAD6mzsBOVrVOC1E3GjAbwYiby-GrV-wTTYGDyRTD33HgHHyj0bCOWYEL1sWWF0BryshD2LCb1hzuFiEQOQLVRILv1RFwIygxjYC5QjXIPEGIFOFEUIImmILHwoMbToKfwYPQGryunzHe1HF9xHxPHHxBG4PYKQLv3QAH2zVSGU3HHyTpUNCQAgRIH+w4DzmvRdE4EhU4FCnikMBMxNlJCMIJhhkTTeSVBVCP3kBPVIDRXQHf0nDbmYG-3TTmG-2UBiAkMfykPHyYLELv0Xz1iM3rS33RBny1G8JklOCP3EFCIKHCKd0iKs2Tn32jHrwknANLHb1k1XgUxki7w0z8jT1Mx0yDm9B8kM1OA4gpDMyqLzG0h93HBnwSLzwH3SL0POlwDVHTHJU4kC1AQcKOgS3wXKMbXqN0Eqw2xQElHMIIWIXYWzQNzsxXVZHrmQAv3HBVBkEhTqEtjN3lFUhA2m3mC4GTBPD2Cgif3jXmFanNSOhnWEAi34GzkLjPSAiiNcIHW-3qziOCMSKX1SO31Tj3wP2yPokYjYGkQOOQDF2D2oPTGELaN+I6OMh9lwiMWbCbDxLNAEDWlbGJIF2Tx9ibDxzNEoFwzxxRFVBBDNCwDNFr2TlbCMT7GjG9GiOaHwlODhJDQRMvhsASKSMQHCLSJTgyIPy5N+K1BMz5PhLF2FI4OBLCLwSMwlJ329EyJlIclwC1Fdz8n5LaiVJFKX1d01PBO1OlNEK1FXwVIFNNJVNFPCOX0tKlM5NtOaG-xkBBDmgHXtPAGNJQCdOkNVOSLwVBO8kDFh3cL22-yM28NQG-0NLDH+kMDlIqP9INOn1lO9KO3gBBFjO-3jMTLkBTKNMVKFLNLVJMzU3VMIQtPdNaDYCSH9ISOyFCBxXXWVAeTxNwmJNFLwWAmtVCB2GogEBAx8mLLHhLI+z8nbJVKzPTIHW02LNLKTIrMbWzOaGXx+L1L+NKyLN9G-1nPuwdJNOrOdJBOkK-zXx12akiJ9VThbNjTLlCFZU9nRnAVpQi1qgamQHEHAWvV4mkE4nJGUHNw32IGfPvO2NEEv2WlFAJmEHIE4mWCpQoLTzqjjTIBQRTRQX1E8OTQtlS0YDOIQTggeFgFiFBHJXNWmMkBQFb1SSki0PzkLhz1VGsNwhsFdzdw91wVwg92QAAFp8JxA6AHh+47jNo5AG5EAeYXCDyYjuBGA1Rv8jkXUFBktv9cI-SVyYiZBKzHSrywyXS8FRKLTcyVLvS1KNKtK0ZwEsAEyDLWBpNvTjKLyQyzKF8l9eKABWSIkGMAmE8PAOXcGoLAXgYgYPVouyX4ggcIZAIwb-BQZYaYEEMecyHKjk9Ii0epL4n5LqE5M5HIZkTYrCS-BWQOCyA3QlJpTKFpVANpf8-uVAVqNAVY7ASgB5EqJIMebIBNR1Z1LMcgfUR4DQZw-cqcLUBc-AZK1K9K2AaYCy6y7NCEuvWXZElEAcCMIAA

**Comment by tadhgmister** [comment]
I think the existing solutions using `[-1,0,1,2,...][depth]` to count down might give better results for fractional or negative indices compared to my solution with the counter. Identifying fractional and negative ones is [technically doable with string templates][1] but it seems highly fragile. To be production ready, the solution in the comment above needs some mechanism to differentiate negative or fractional numbers from sufficiently large numbers to flatten the whole array. (also probably more thorough testing) The code I link to in this comment does do this but I'm not sure it is a good fit for the standard library.

[1]: https://www.typescriptlang.org/play/?noLib=true&target=6#code/PQKhAICcFMBcFdIDsDO5YAtrgIYCMUB7AG3lmwDcdTtCAzdLcAcwEsLolwl4BbPaJHDFW5SNXAhgAKFgBPAA7YAggCEAygB5p4cADlw0AB7kkAEzQ9+g6QD5wAXn2GTnC7iRzdwYC4U5zcEIOIWpicF5oa0g0ekZseCRWQiQdXXAAfnAACgADABIAbz0AX1yXU3dcgFoi1iQ6QXAABQq3Sz4BSDLMlvAALn0ASjTBpGgQ8B9wAGMcFGwAdywYcFyDY0q0ALly1jQ6agWc8cmYHBmsMxHpUAgYBGQ0WEh4bFYGTGw2Di4rLvA+3AZmgdHqomgxC842YOFg7GgABpwDoIIdiMcPoDYIC0ApCCh4RxkYQhLk8IQSNAAnsGKJZoR4MQzOABIZRCsAHQoyQyeRKcAASXUAH09ABRADiygAKoKAGrikWCvQAEXFAA1tLoNq5zB1onZHNxOk1Nu1nGldFkKVSAjyBs5zfrwGotHp7Fl0cdBi83rcwDyIILPkwfpwTdFcR5AUhyMwmpggQ9EKhwAByb3QdOooWh74Iv6moRApCEHH2+rxxMYZNwVNodN+7O5kPYhlMlls6AcmtA8ON23EalIHOSIPgABKkNY0FiXEJkHqzHQUQUxDh2BEYgkAgwOAoyUQ4DopMjXW2gWFYqlsoVSpV6o1qL5imw14AYpPlABhOUAeT0ZU1U1bUnT1dx-hsewnCgoRnXcAwbUpYd7VGcCtg8LwrV6Apih6BC0Dw+pGiEABZHAV0I89BBKaA6gaJpxSMBQ2hdOCehw60hVFCVpTlRVgKfTRmIUewcMGPDSnKajiMYoRBTjNjIOLEpOQY0jwFVaAZmUg0ujKLjembKZfDLchGCBRdWBmWAoSgGc5yCLgAEZOQABlZHskBXGBFiXWBTFwNAXIkk8jmwdDThsaRZDfcBxQAR00ZRkVUPTXRg8AAG1lAAXQy7LVAKrJ0zkOd00ddMyxzJAcEiFB-BmbA+LvQSfwACXFH8AGl1HAQppBmFJCW4aBmBcwYks0a9WoEh8QK1aoXNsRFmyysqKoAbiGkacRhAAmKbktm295qE0DlugFz3Pc1b1uNTaUHTHbhtQfbxoAZmOmbeLO+8LqWgBWTkgfu15oA28rnp23b3vAfEUEmhKTr+-iAcfUC7sRLMoe20zY3iXBIHELxhx8zA4dGxGjpR36b3RwTMa1dyAB8VpxiK8ZhgmKUwYnScMYdIjjFAqZxRHvrp07GYW4SXOu27Vtxx7oZegna2YFZGHtL4BZwMnOGYSm3upgkABYfpltq5ax9nsZVpwnvV6Y+YwQmAhZRkcTiHASYN2LTZxN3kem63zuZzQ2eW1ahxHbmXvF1lywwWmw7Rm3Ac0NygdZ5bsbjgIE9hoOT3EGZQ9RhnM8jnPla51X8aTuhy7Tqu5oxxbs+gaoDodhunbVku9sjEUditjOI67nZY5Q+PG5hpOrBFKwJ+rqfhLg2e7SQYvpBKaQ6oaprsC-X8RU67q+oGpOq0r+mz7-QVAKzjnwoxSGF8T0uqzbh-vyfi-SOvdVrvwWHvH+cYpbpxFI-ACQFI5s0+qAx2GYh63zjJbaWoo4HPwQV3Nyn1roHRQQPNB20MGwCBmvXBQCCHuQ8qQj+ECR5VgAGw0IAfA1+t0e5v1Qc7Ye8MW4XHvp+LheCs4MKBuAVm4A3IAHZQEPUHhQ0uIiZh-3Eb+bhtc+HKIhiw4R5doFV1ofg+Wfce4kORCo8hi91HlywTA8xWdlqcnNkQt+djBGBxHrwVgRgxE4IkXQ4SbNpGgMLrvL+r1-GBK0SEnRkjI5b2RNEoxo0AlGFMf-ZJYTQIz3SXPIusT95AA
